Feat/label merge v2

jh3y opened the pull request:
- Different approach to #13 
- Implements custom React hook so that users can use a style and label value provided from a hook
- Requires storybook upgrade etc.

tajo commented on examples/LabeledMerge.tsx:
```diff
@@ -0,0 +1,142 @@
+        borderRadius: '4px',
+        backgroundColor: '#548BF4',
+        whiteSpace: 'nowrap',
+        ...(style as {})
```
```suggestion
        ...(style as React.CSSProperties)
```

jh3y replied:
Thanks 👍

tajo commented on examples/LabeledMerge.tsx:
```diff
@@ -0,0 +1,142 @@
+    overlaps: any[];
+  } = {
+    values: [25, 50, 70, 80, 90],
+    overlaps: []
```
seems to be unused

tajo commented on src/types.ts:
```diff
@@ -48,6 +48,11 @@ export interface ITrackProps {
 
+export interface IRangeRef {
+  getOffsets: () => any;
+  getThumbs: () => any;
```
```
getThumbs: () => Element[];
```

?

tajo commented on src/utils.ts:
```diff
@@ -142,3 +143,106 @@ export function voidFn() {}
+ * @param separator - string to separate thumb values
+ */
+export const useThumbOverlap = (
+  rangeRef: IRangeRef | undefined,
```
```suggestion
  rangeRef?: IRangeRef,
```

tajo replied:
Also, if this can be undefined, should it be the last param?

jh3y replied:
The hook doesn't really work without it but the ref value can be `null` the first time the hook is run. The hook won't do anything without the ref value.

tajo commented on src/types.ts:
```diff
@@ -48,6 +48,11 @@ export interface ITrackProps {
 }
 
+export interface IRangeRef {
+  getOffsets: () => any;
```
Can the return be typed?

tajo commented on package.json:
```diff
@@ -38,17 +38,18 @@
-    "@storybook/addon-options": "4.2.0-alpha.8",
-    "@storybook/addons": "4.2.0-alpha.8",
-    "@storybook/react": "4.2.0-alpha.8",
+    "babel-loader": "^8.0.6",
```
what's up with `babel`?

jh3y replied:
It's required for `storybook` v5. Without it, `storybook` won't build/load. They added something to their docs about it becoming a dependency.

tajo commented on src/utils.ts:
```diff
@@ -142,3 +143,106 @@ export function voidFn() {}
+  index: number,
+  separator = '-',
+) => {
+  const [style, setStyle] = useState<{ [key: string]: any }>({});
```
`React.CSSProperties`

tajo commented on src/utils.ts:
```diff
@@ -142,3 +143,106 @@ export function voidFn() {}
+      let labelValue = values[index].toFixed(1);
+      const offsets = rangeRef.getOffsets();
+      const THUMB_SIZE = rangeRef.getThumbs()[0].offsetWidth;
+      const newStyle: { [key: string]: any } = {};
```
`React.CSSProperties`

tajo commented on src/utils.ts:
```diff
@@ -142,3 +143,106 @@ export function voidFn() {}
+      let overlaps = [];
+      let labelValue = values[index].toFixed(1);
+      const offsets = rangeRef.getOffsets();
+      const THUMB_SIZE = rangeRef.getThumbs()[0].offsetWidth;
```
I would check if `rangeRef.getThumbs().length > 0`. We throw a warning already but it would nice to prevent the crash.

tajo commented on src/utils.ts:
```diff
@@ -142,3 +143,106 @@ export function voidFn() {}
+
+  useEffect(() => {
+    if (rangeRef) {
+      let overlaps = [];
```
`Number[]` ?

tajo commented on src/utils.ts:
```diff
@@ -142,3 +143,106 @@ export function voidFn() {}
+              (currentX + THUMB_SIZE >= siblingX &&
+                currentX + THUMB_SIZE <= siblingX + THUMB_SIZE))
+          ) {
+            let overlapIndex;
```
```suggestion
            let overlapIndex?: number;
```

tajo commented on examples/LabeledMerge.tsx:
```diff
@@ -0,0 +1,142 @@
+        <Range
+          allowOverlap
+          values={this.state.values}
+          ref={r => (this.RANGE = r)}
```
`this.rangeRef` might be better naming? Also, `React.createRef()` seems to be the preferred way over callbacks since React `16.3`.

tajo commented:
**I really like the hook, happy to see that we can abstract it!** 

Some notes:

- Can you type all the `any`s? Especially the variable `overlaps` is critical to understand what's happening in the code.
- It's quite hard to follow the hook's code. Some things that could help to make it more readable:

  - Replace `for` with `forEach` to reduce usage of one-letter indexes
  - type the reduce functions
  - don't use one-letter variables

- Are e2e screenshots updated? What happened?
- Can we add e2e tests for this feature?
- Some parts of the hook could be split into a different function, like the part where you build the `overlaps`. 
- Could you add more inline comments explaining the whole algorithm (especially what output you need). I have only a vague idea what it does. There might be some more efficient solution, it's similar to dynamic programming problems. No need to over-optimize it now but for the future reference it would be nice to document what's going on.
- You'll need to update `/flowtypes/utils.js.flow`. I recommend to `yarn link` this in `baseui` and implement the functionality there to test the flow types.

jh3y commented:
Hey @tajo 👋 
Updates made:
- [x] Typed any existing `any`s 👍 
- [x] Inline commented the hook code + refactored the overlap building. It's much smaller and uses recursion 💪 
- [x] `e2e` snapshots updated with no problems
- [x] Updated `flow` declaration for use in `baseui`. However, `yarn link` doesn't seem to be working as expected for me.
- [x] Added `e2e` tests that test the different overlaps and which labels should be hidden and their values. However, these tests pass on `test:e2e:dev` but fail on `test:e2e`. Is there something I'm missing? I've watched `Chromium` do it's thing and all looks good but when running `test:e2e` it fails on each snapshot. It's like the viewport dimensions are different between the environments.

gergelyke commented on examples/LabeledMerge.tsx:
```diff
@@ -0,0 +1,139 @@
+  const [labelValue, style] = useThumbOverlap(rangeRef, values, index);
+  return (
+    <div
+      style={{
```
in puppeteer, we want to be able to target these elements, so let's add an identifier:

```
data-label={index}
```

gergelyke commented on e2e/merged.test.ts:
```diff
@@ -0,0 +1,65 @@
+  await page.mouse.up();
+  await untrackMouse(page);
+  const output = await page.$('[draggable]:nth-of-type(1) div');
+  const hidden = await page.$('[draggable]:nth-of-type(2) div');
```
```
  await page.waitForSelector('[data-label="1"]', {
    hidden: true
  });
```

gergelyke commented on e2e/merged.test.ts:
```diff
@@ -0,0 +1,65 @@
+  await page.mouse.up();
+  await untrackMouse(page);
+  const output = await page.$('[draggable]:nth-of-type(1) div');
+  const hidden = await page.$('[draggable]:nth-of-type(2) div');
```
```
  await page.waitForSelector('[data-label="1"]', {
    hidden: true
  });

  await page.waitForSelector('[data-label="2"]', {
    hidden: true
  });

```

gergelyke commented on e2e/merged.test.ts:
```diff
@@ -0,0 +1,65 @@
+  await page.mouse.up();
+  await untrackMouse(page);
+  const output = await page.$('[draggable]:nth-of-type(3) div');
+  const hidden = await page.$('[draggable]:nth-of-type(1) div');
```
```
  await page.waitForSelector('[data-label="0"]', {
    hidden: true
  });

  await page.waitForSelector('[data-label="1"]', {
    hidden: true
  });
```

jh3y commented:
Thanks so much for taking a look at this @gergelyke 👍 
Tests updated and passing 🙌

gergelyke commented:
@tajo @jh3y can we simply remove the visual diffs if they seem to be unreliable? puppeteer tests the functionality anyways
